keps/sig-node/3288-separate-stdout-from-stderr/README.md

```diff
@@ -0,0 +1,817 @@
+
+### Goals
+
+- Enable api-server to return certain log stream of a container
+- Enable users to fetch certain log stream of a container
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add a new field `Stream` to `PodLogOptions` to allow users to indicate which log stream
+they want to retrieve. To maintain backward compatibility, if this field is not set,
+the combined stdout and stderr from the container will be returned to users.
+
+### User Stories
+
+Bob runs a service "foo" that continuously prints informational messages to stdout,
+and prints warnings/errors to stderr. When the service "foo" is misbehaving, Bob wants
+to know what is going on as soon as possible, so he checks the logs of the service "foo",
+but the error message is not easy to notice because the apiserver always returns the combined
+stdout and stderr from the container.
+
+This problem could be resolved if Bob is able to specify that he only wants stderr.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+
+## Design Details
+
+### Changes of kube-apiserver
+
+Add a new field `Stream` to `k8s.io/kubernetes/pkg/apis/core.PodLogOptions`:
+
+```go
+// PodLogOptions is the query options for a Pod's logs REST call
+type PodLogOptions struct {
+...
+    // If set, return the given log stream of the container.
+	// Otherwise, the combined stdout and stderr from the container is returned.
+	// Available values are: "stdout", "stderr", "" (empty string).
+    Stream string
+}
+```
+
+When users want to query certain stream from container, they need to add a new query named `stream`
+to the URL, i.e. `/api/v1/namespaces/default/pods/foo/log?stream=stderr&container=nginx`.
+Then the kube-apiserver is able to know the desired `stream` and passes it to the kubelet.
+
+To tell kubelet which stream to return, we need to update the `LogLocation` function to make it be aware of
+the new parameter:
+
+```go
+// LogLocation returns the log URL for a pod container. If opts.Container is blank
+// and only one container is present in the pod, that container is used.
+func LogLocation(
+	ctx context.Context, getter ResourceGetter,
+	connInfo client.ConnectionInfoGetter,
+	name string,
+	opts *api.PodLogOptions,
+) (*url.URL, http.RoundTripper, error) {
+	...
+    params := url.Values{}
+    // validate stream
+    switch opts.Stream {
+        case "stdout", "stderr", "":
+        default:
+            return nil, nil, errors.NewBadRequest(fmt.Sprintf("invalid container log stream %s", opts.Stream))
+    }		
+    params.Add("stream", opts.Stream)
+	...
+}
+```
+
+### Changes of kubelet
```

Comment:
> will there be any issues with TailLines log option?

If `TailLines` is set, only the last N lines in the log file whose stream match the one user specified would be returned.

> What will be the behavior of this log option when only one stream was specified?

Then only that log stream from the beginning  would be returned.

Reply:
I haven't looked into code, I just anticipate possible additional changes needed to ensure that the counting of TailLines happens properly and only logs from a specified stream were counted. Definitely add this to the test plan section

Reply:
> I just anticipate possible additional changes needed to ensure that the counting of TailLines happens properly and only logs from a specified stream were counted.

Actually my original thought is slightly different, I expected kubelet still examine the last `TailLines` of the log file, but only returns lines that matches the stream user specified. 

I also understand my original thought is easy to implement, though it might confuse users. I agree that the most intutive way is only counting logs from a specified stream,  and it would require extra work to be done in kubelet.

Reply:
+1 on counting logs from a specified stream.